python_modules/libraries/dagster-dg/dagster_dg/config.py

```diff
@@ -390,11 +402,54 @@ def is_project_file_config(config: "DgFileConfig") -> TypeGuard[DgProjectFileCon
 DgFileConfig: TypeAlias = Union[DgWorkspaceFileConfig, DgProjectFileConfig]
 
 
+def is_dg_specific_config_file(path: Path) -> bool:
+    """Check if the file is a dg-specific toml file."""
+    return path.name == "dg.toml" or path.name == ".dg.toml"
+
+
+@contextmanager
+def modify_dg_toml_config(path: Path) -> Iterator[Union[tomlkit.TOMLDocument, tomlkit.items.Table]]:
+    """Modify a TOML file as a tomlkit.TOMLDocument, preserving comments and formatting."""
+    with modify_toml(path) as toml:
+        if is_dg_specific_config_file(path):
+            yield toml
+        elif not has_toml_node(toml, ("tool", "dg")):
+            raise KeyError(
+                "TOML file does not have a tool.dg section. This is required for pyproject.toml files."
+            )
+        else:
+            yield get_toml_node(toml, ("tool", "dg"), tomlkit.items.Table)
+
+
+@contextmanager
+def modify_dg_toml_config_as_dict(path: Path) -> Iterator[dict[str, Any]]:
+    """Modify a TOML file as a plain python dict, destroying comments and formatting.
+    This will take account of the filename and yield only the dg config part. This is the root node
+    for dg.toml files and the tool.dg section otherwise.
+    """
+    with modify_toml_as_dict(path) as toml_dict:
+        if is_dg_specific_config_file(path):
+            yield toml_dict
+        elif not has_toml_node(toml_dict, ("tool", "dg")):
+            raise KeyError(
+                "TOML file does not have a tool.dg section. This is required for pyproject.toml files."
+            )
+        else:
+            yield get_toml_node(toml_dict, ("tool", "dg"), dict)
+
+
 def has_dg_file_config(
```

Comment:
this is new to me, but it seems weird to pass in a the `predicate` here rather than doing it at callsites, since it is at the very end of the function

Reply:
The reason I implemented it this way is to consolidate the defensive programming into one call instead of needing extra conditional logic wherever we use this to make sure that `tool.dg` exists. It makes the callsites simpler and more readable.

Reply:
kk i don't feel too strongly